### ISSUE TITLE
Plugins with new dependencies fails to upgrade

### ISSUE THREAD

**Issue by JohnXLivingston** [Chocobozzz/PeerTube#4593]
**Describe the current behavior**

This is not the first time I have such an issue, and I don't understand exactly when and why it happens. Here is the best description I can make of this issue.

- I have the plugin peertube-plugins-livechat in version 4.0.3
- I have several other themes and plugins
- my Peertube server is a production server
- I try to update the plugin to v5.0.1 (that I just released).
- Note: there is a new npm dependency in v5.0.1: 'validate-color'
- After clicking on «upgrade to v5.0.1», the upgrade fails (see the logs at the end of the bug report)
- No mater what I do (re-install v4.0.3, upgrade, ...), it always fails
- There is no `validate-color` in /var/www/peertube/storage/plugins/node_modules
- In order to achieve the upgrade, I have to restart Peertube. Then the «upgrade to v5.0.1» button works.

Sorry that I cannot be more specific on what happens.

**Describe the expected behavior**

Plugins should upgrade successfully by installing new dependencies.

**Additional information**

* PeerTube instance:
  * URL: https://www.yiny.org
  * version: 3.4.1
  * NodeJS version: v14.18.1 (yarn 1.22.15)

* Link to server log if relevant (`journalctl` or `/var/www/peertube/storage/logs/`):

```
error[01/12/2021, 15:16:26] Cannot install plugin peertube-plugin-livechat, removing it...

{
  "err": {
    "stack": "Error: Cannot find module 'validate-color'\nRequire stack:\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server.js\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)\n    at Function.Module._resolveFilename (/var/www/peertube/versions/peertube-v3.4.1/node_modules/tsconfig-paths/lib/register.js:75:40)\n    at Function.Module._load (internal/modules/cjs/loader.js:746:27)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js:4:23)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js:28:22)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js:25:19)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js:4:15)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)",
    "message": "Cannot find module 'validate-color'\nRequire stack:\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server.js",
    "code": "MODULE_NOT_FOUND",
    "requireStack": [
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server.js"
    ]
  }
}


```

**Comment by Chocobozzz** [comment]
Hello,

Please paste logs above your error, when peertube upgrades/installs the plugin.

**Comment by JohnXLivingston** [comment]
Here there are. I'm not sure exactly when it begins, because I make multiple attemps. And I only have info level, as it is a production server.

```json
{
  "tags": [
    "http"
  ],
  "level": "info",
  "message": "2a01:: - - [01/Dec/2021:14:17:53 +0000] \"GET /api/v1/plugins?start=0&count=10&sort=name&pluginType=1 HTTP/1.0\" 200 2118 \"https://www.yiny.org/admin/plugins/list-installed\" \"Firefox\"",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:53.629Z"
}
{
  "level": "info",
  "message": "Updating plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.153Z"
}
{
  "level": "info",
  "message": "Unregister plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.162Z"
}
{
  "message": "Checking if Prosody is running, and shutting it down if so",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.165Z"
}
{
  "level": "info",
  "message": "Deleted locale ca-ES of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.163Z"
}
{
  "level": "info",
  "message": "Deleted locale fr-FR of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.163Z"
}
{
  "level": "info",
  "message": "Deleted locale eo of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.164Z"
}
{
  "level": "info",
  "message": "Deleted locale eu-ES of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.164Z"
}
{
  "level": "info",
  "message": "Deleted locale pl-PL of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.164Z"
}
{
  "level": "info",
  "message": "Deleted locale it-IT of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.165Z"
}
{
  "message": "Stoping Prosody log rotation",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.166Z"
}
{
  "message": "Prosody process exited with code 0",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.252Z"
}
{
  "message": "Prosody process closed all stdio with code 0",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.253Z"
}
{
  "message": "ProsodyCtl command returned: WARNING: Use of prosodyctl start/stop/restart/reload is not recommended\n         if Prosody is managed by an init system - use that directly instead.\n         e.g. systemctl stop prosody\n\nStopped\n",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.727Z"
}
{
  "message": "Regenerating registered plugin CSS to global file.",
  "level": "info",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.741Z"
}
{
  "message": "Changing http-bind port for null",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.728Z"
}
{
  "message": "Unloading module /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js...",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.729Z"
}
{
  "message": "Successfully unloaded the module /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js",
  "level": "info",
  "label": "www.yiny.org:443 peertube-plugin-livechat",
  "timestamp": "2021-12-01T14:17:58.740Z"
}
{
  "message": "Cleaning HTML cache.",
  "level": "info",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.742Z"
}
{
  "message": "Cleaning HTML cache.",
  "level": "info",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.747Z"
}
{
  "message": "Cleaning HTML cache.",
  "level": "info",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.747Z"
}
{
  "level": "info",
  "message": "Installing plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:17:58.748Z"
}
{
  "level": "info",
  "message": "Successful installation of plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:18:04.494Z"
}
{
  "level": "info",
  "message": "Registering plugin or theme peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:18:04.496Z"
}
{
  "level": "error",
  "message": "Cannot install plugin peertube-plugin-livechat, removing it...",
  "label": "www.yiny.org:443",
  "err": {
    "stack": "Error: Cannot find module 'validate-color'\nRequire stack:\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server.js\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)\n    at Function.Module._resolveFilename (/var/www/peertube/versions/peertube-v3.4.1/node_modules/tsconfig-paths/lib/register.js:75:40)\n    at Function.Module._load (internal/modules/cjs/loader.js:746:27)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js:4:23)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js:28:22)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js:25:19)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js:4:15)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)",
    "message": "Cannot find module 'validate-color'\nRequire stack:\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server.js",
    "code": "MODULE_NOT_FOUND",
    "requireStack": [
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server.js"
    ]
  },
  "timestamp": "2021-12-01T14:18:04.536Z"
}
{
  "level": "info",
  "message": "Uninstalling plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:18:04.539Z"
}
{
  "level": "info",
  "message": "Unregister plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:18:04.539Z"
}
{
  "level": "warn",
  "message": "Cannot unregister plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "err": {
    "stack": "Error: Unknown plugin peertube-plugin-livechat to unregister\n    at PluginManager.<anonymous> (/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js:181:23)\n    at Generator.next (<anonymous>)\n    at /var/www/peertube/versions/peertube-v3.4.1/node_modules/tslib/tslib.js:117:75\n    at new Promise (<anonymous>)\n    at Object.__awaiter (/var/www/peertube/versions/peertube-v3.4.1/node_modules/tslib/tslib.js:113:16)\n    at PluginManager.unregister (/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js:177:24)\n    at PluginManager.<anonymous> (/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js:261:28)\n    at Generator.next (<anonymous>)\n    at /var/www/peertube/versions/peertube-v3.4.1/node_modules/tslib/tslib.js:117:75\n    at new Promise (<anonymous>)\n    at Object.__awaiter (/var/www/peertube/versions/peertube-v3.4.1/node_modules/tslib/tslib.js:113:16)\n    at PluginManager.uninstall (/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js:258:24)\n    at PluginManager.<anonymous> (/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js:228:32)\n    at Generator.throw (<anonymous>)\n    at rejected (/var/www/peertube/versions/peertube-v3.4.1/node_modules/tslib/tslib.js:115:69)",
    "message": "Unknown plugin peertube-plugin-livechat to unregister"
  },
  "timestamp": "2021-12-01T14:18:04.540Z"
}
{
  "level": "info",
  "message": "Plugin peertube-plugin-livechat uninstalled.",
  "label": "www.yiny.org:443",
  "timestamp": "2021-12-01T14:18:10.834Z"
}
{
  "level": "warn",
  "message": "Cannot update plugin peertube-plugin-livechat.",
  "label": "www.yiny.org:443",
  "err": {
    "stack": "Error: Cannot find module 'validate-color'\nRequire stack:\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server.js\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)\n    at Function.Module._resolveFilename (/var/www/peertube/versions/peertube-v3.4.1/node_modules/tsconfig-paths/lib/register.js:75:40)\n    at Function.Module._load (internal/modules/cjs/loader.js:746:27)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js:4:23)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js:28:22)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js:25:19)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n    at Module.load (internal/modules/cjs/loader.js:950:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n    at Module.require (internal/modules/cjs/loader.js:974:19)\n    at require (internal/modules/cjs/helpers.js:93:18)\n    at Object.<anonymous> (/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js:4:15)\n    at Module._compile (internal/modules/cjs/loader.js:1085:14)",
    "message": "Cannot find module 'validate-color'\nRequire stack:\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js\n- /var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js\n- /var/www/peertube/versions/peertube-v3.4.1/dist/server.js",
    "code": "MODULE_NOT_FOUND",
    "requireStack": [
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/shared/lib/autocolors.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/routers/webchat.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/prosody/ctl.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/lib/settings.js",
      "/var/www/peertube/storage/plugins/node_modules/peertube-plugin-livechat/dist/server/main.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/plugin-manager.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/lib/plugins/theme-utils.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/models/user/user.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server/initializers/checker-after-init.js",
      "/var/www/peertube/versions/peertube-v3.4.1/dist/server.js"
    ]
  },
  "timestamp": "2021-12-01T14:18:10.836Z"
}

```



For the record, here is the package.json file on my server, after succesfully upgrading the plugin:
```json
{
  "dependencies": {
    "peertube-plugin-georegister": "0.0.5",
    "peertube-plugin-livechat": "5.0.1",
    "peertube-plugin-simplelogo": "0.0.5",
    "peertube-theme-dark": "2.2.5"
  }
}
```

**Comment by Chocobozzz** [comment]
It's weird I can't reproduce in latest develop :thinking: 

Have you heard other admins having this issue?

**Comment by JohnXLivingston** [comment]
> It's weird I can't reproduce in latest develop thinking
> 
> Have you heard other admins having this issue?

No. But it's not the first time I have this bug. The first 2 or 3 times, I thought I had done a bad manipulation. But yesterday I was sure of my move. And it happened several times, until I restarted the server.

I wonder if it's not related to the fact that I installed themes between the installation and the upgrade, without restarting the server. 

I'll try to do some more tests. 

And if I get any feedback from other people who had this happen, I'll let you know.

**Comment by JohnXLivingston** [comment]
Is there an easy way to change plugin version backward and forward to test? Using npm for downloading the package, to be sure to test the same bug?

**Comment by Chocobozzz** [comment]
You can use https://docs.joinpeertube.org/maintain-tools?id=plugin-installuninstall to install a specific version of the plugin. Then restart peertube and you should be able to upgrade your plugin

**Comment by JohnXLivingston** [comment]
I made some more tests. It seems that it won't happen on «normal» environments. It seems related to the fact that I somehow installed the plugin from a local folder before trying to upgrade it with npm.

Here is a way to reproduce (v4.0.0-RC1, but I think it would be the same with 3.4.1):

1. Download an old version of https://github.com/JohnXLivingston/peertube-plugin-livechat (for example tag 4.0.3) in /tmp/peertube-plugin-livechat
2. Setup the peertube CLI
2. Restart the peertube server
3. Install the old version of the plugin from disk with `node ./dist/server/tools/peertube.js plugins install --path /tmp/peertube-plugin-livechat/` (in the folder where the CLI was build)
4. Go the the plugin page in the Peertube interface
5. If the button «upgrade to 5.0.1» is not here, do the following in psql: `update "plugin" set "latestVersion" = '5.0.1' where "plugin"."name" = 'livechat';`
6. Click on «upgrade to 5.0.1». It should fail.
7. Restart Peertube
8. Click again. It should work.

**Comment by Chocobozzz** [comment]
Okay thanks, I think I managed to fix this bug in https://github.com/Chocobozzz/PeerTube/commit/ca87d95bcbfec9241e5840267862bace8b0197fa

It won't be included in v4 since it's not a trivial change.

**Comment by JohnXLivingston** [comment]
> Okay thanks, I think I managed to fix this bug in [ca87d95](https://github.com/Chocobozzz/PeerTube/commit/ca87d95bcbfec9241e5840267862bace8b0197fa)
> 
> It won't be included in v4 since it's not a trivial change.

Great!
Didn't you forget to uncomment these lines?
https://github.com/Chocobozzz/PeerTube/blob/ca87d95bcbfec9241e5840267862bace8b0197fa/server/lib/plugins/plugin-manager.ts#L315
https://github.com/Chocobozzz/PeerTube/blob/ca87d95bcbfec9241e5840267862bace8b0197fa/server/lib/plugins/plugin-manager.ts#L320

**Comment by JohnXLivingston** [comment]
I just noticed this line:
https://github.com/Chocobozzz/PeerTube/blob/ca87d95bcbfec9241e5840267862bace8b0197fa/server/lib/plugins/plugin-manager.ts#L289

I think that npmName should be read in the plugin/theme package.json file. Otherwise, when installing with --path, it can be wrong (if the folder has a different name).
Now that I have seen this line, I think I had some issues in the past that can have been related...
